### PR TITLE
bucket versioning #1 : prerequisites

### PR DESCRIPTION
**Pull request by ttvi**
Step 1 (this PR): S3 to support versioning for buckets.
Step 2: S3 to provide routes for interacting with Metadata.
Step 3: S3 to provide versioning support for each operation.
Step 4: MetaData to support versioning.

### CODE REVIEW

**Comment by ttvi** [comment]
@ironman-machine let's go beta (then ramen for lunch)

SCALITY_FEDERATION_BRANCH=FT/metadata-vault-disable-versioning
SCALITY_METADATA_BRANCH=exp/geobucket_versioning
SCALITY_VAULT_BRANCH=master

File: lib/api/bucketGet.js [hunk 1]
```diff
@@ -29,6 +31,177 @@ import { errors } from 'arsenal';
...
+    list.CommonPrefixes.forEach(item => {
+        xml.push(
+            `<CommonPrefixes><Prefix>${item}</Prefix></CommonPrefixes>`
+        );
+    });
+    xml.push('</ListBucketResult>');
+    return xml.join('');
+}
+
+// Sample XML response for versioning:
```
Comment (LaurenSpiegel) [inline lib/api/bucketGet.js]: Sample XML response for GET Bucket Object Versions API

File: lib/api/bucketGet.js [hunk 2]
```diff
@@ -63,9 +238,13 @@ export default function bucketGet(authInfo, request, log, callback) {
         delimiter: params.delimiter,
         marker: params.marker,
         prefix: params.prefix,
+
     };
+    if (versions) {
+        listParams.versions = versions;
+    }
```
Comment (LaurenSpiegel) [inline lib/api/bucketGet.js]: I don't think you use listParams.versions so don't need to do this. If you do need it, it would be clearer to just set listParams.versions to true.

File: lib/api/bucketGet.js [hunk 3]
```diff
@@ -29,6 +31,177 @@ import { errors } from 'arsenal';
...
+            xml.push(`<${param.tag}>${param.value}</${param.tag}>`);
+        } else {
+            xml.push(`<${param.tag}/>`);
+        }
+    });
+
+    list.Contents.forEach(item => {
+        const key = encoding === 'url' ?
+            querystring.escape(item.key) : escapeForXML(item.key);
+        const v = VSUtils.decodeVersion(item.value);
```
Comment (LaurenSpiegel) [inline lib/api/bucketGet.js]: Since this is a json.parse it should be wrapped in a try/catch here or in the utils function itself.

File: lib/api/bucketGet.js [hunk 4]
```diff
@@ -29,6 +31,177 @@ import { errors } from 'arsenal';
...
+        const size = v['content-length'];
+        const ETag = v['content-md5'];
+        const lastModified = v['last-modified'];
+        const owner = {
+            DisplayName: v['owner-display-name'],
+            ID: v['owner-id'],
+        };
+        const storageClass = v['x-amz-storage-class'];
+
+        const versionId = VSUtils.getts(v);
```
Comment (LaurenSpiegel) [inline lib/api/bucketGet.js]: If you are using getters on some attributes might as well be consistent and do it for all.

File: lib/api/bucketGet.js [hunk 5]
```diff
@@ -29,6 +31,177 @@ import { errors } from 'arsenal';
...
+        const isLatest = false;
```
Comment (LaurenSpiegel) [inline lib/api/bucketGet.js]: How do you know this?  Is this a todo to make dynamic later?

File: lib/api/bucketGet.js [hunk 6]
```diff
@@ -29,6 +31,177 @@ import { errors } from 'arsenal';
...
+                '</DeleteMarker>'
+            );
+        } else {
+            xml.push(
+                '<Version>',
+                `<Key>${key}</Key>`,
+                `<VersionId>${versionId}</VersionId>`,
+                `<IsLatest>${isLatest}</IsLatest>`,
+                `<LastModified>${lastModified}</LastModified>`,
+                `<ETag>${ETag}</ETag>`,
```
Comment (LaurenSpiegel) [inline lib/api/bucketGet.js]: Your code made me realize that the etag listing should appear like this:
`<ETag>&quot;${ETag}&quot;</ETag>`

It should be this way in both listings.

File: lib/api/objectGet.js [hunk 7]
```diff
@@ -36,6 +39,10 @@ function objectGet(authInfo, request, log, callback) {
                 log.debug('error processing request', { error });
                 return callback(error);
             }
+            // return ObjNotFound if the version is a delete marker
+            if (VSUtils.getdm(objMD)) {
+                return callback(errors.ObjNotFound);
```
Comment (LaurenSpiegel) [inline lib/api/objectGet.js]: I think the correct error is NoSuchKey.  See s3 errors: http://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html

Also, it looks like you need to add a header to the response:
"If the current version of the object is a delete marker, Amazon S3 behaves as if the object was deleted and includes x-amz-delete-marker: true in the response."

File: lib/api/objectPut.js [hunk 8]
```diff
@@ -46,6 +46,10 @@ function _storeIt(bucketName, objectKey, objMD, authInfo, canonicalID,
         dataToDelete = Array.isArray(objMD.location) ?
             objMD.location : [objMD.location];
     }
+    // not deleting old data in the case of versioning
+    if (objMD && objMD['x-amz-version-id'] !== 'null') {
+        dataToDelete = undefined;
+    }
```
Comment (LaurenSpiegel) [inline lib/api/objectPut.js]: To be backwards compatible, shouldn't you do:
`if(objMD && objMD['x-amz-version-id'] && objMD['x-amz-version-id'] !== 'null')`

This assumes that you will be setting the objMD['x-amz-version-id'] to null for all non-versioned objects.

File: lib/api/objectPut.js [hunk 9]
```diff
@@ -10,14 +10,14 @@ import { logger } from '../utilities/logger';
 function _storeInMDandDeleteData(bucketName, objMD, dataGetInfo,
     metadataStoreParams, dataToDelete, deleteLog, callback) {
     services.metadataStoreObject(bucketName, objMD, dataGetInfo,
-        metadataStoreParams, (err, contentMD5) => {
+        metadataStoreParams, (err, value) => {
```
Comment (LaurenSpiegel) [inline lib/api/objectPut.js]: what is being sent back now?

File: lib/routes/routePUT.js [hunk 10]
```diff
@@ -104,6 +104,11 @@ export default function routePUT(request, response, log) {
                 }
                 // ETag's hex should always be enclosed in quotes
                 const resMetaHeaders = { ETag: `"${request.calculatedHash}"` };
+                if (data) {
+                    const _data = typeof data === 'object' ?
```
Comment (LaurenSpiegel) [inline lib/routes/routePUT.js]: Why is this sometimes an object?

File: lib/routes/routePUT.js [hunk 11]
```diff
@@ -104,6 +104,11 @@ export default function routePUT(request, response, log) {
...
+                        data : JSON.parse(data);
```
Comment (LaurenSpiegel) [inline lib/routes/routePUT.js]: should put in a try/catch

File: lib/services.js [hunk 12]
```diff
@@ -258,24 +260,30 @@ export default {
                     return cb(err);
                 }
                 omVal.acl = parsedACL;
-                metadata.putObjectMD(bucketName, objectKey, omVal, log, err => {
+                metadata.putObjectMD(bucketName, objectKey, omVal, log,
```
Comment (LaurenSpiegel) [inline lib/services.js]: shouldn't you be updating the omVal['x-amz-server-version-id'] above?

File: lib/api/objectPut.js (hunk 8 above)
Comment (LaurenSpiegel) [inline lib/api/objectPut.js]: I see now that objMD['x-amz-version-id'] was set to 'null' in current implementation so ok on backwards compatibility but shouldn't we still be able to delete data if versioning is not enabled?

File: lib/services.js [hunk 13]
```diff
@@ -258,24 +260,30 @@ export default {
...
+                    (err, data) => {
+                        if (err) {
+                            log.error('error from metadata', { error: err });
+                            return cb(err);
+                        }
+                        log.trace('object successfully stored in metadata');
+                        const value = data ? JSON.parse(data) : {};
```
Comment (LaurenSpiegel) [inline lib/services.js]: Could we have a class for the data coming back with a deserialize method that is called in the wrapper?

File: lib/services.js [hunk 14]
```diff
@@ -288,18 +296,21 @@ export default {
...
      * @return {undefined}
      */
-    deleteObject(bucketName, objectMD, responseMDHeaders, objectKey, log, cb) {
+    deleteObject(bucketName, objectMD, responseMDHeaders, objectKey, log, cb,
+        query) {
         log.trace('deleting object from bucket');
         assert.strictEqual(typeof bucketName, 'string');
         assert.strictEqual(typeof objectMD, 'object');
-        if (objectMD['x-amz-version-id'] === 'null') {
+        if (objectMD['x-amz-version-id'] === 'null' || query.versionId) {
```
Comment (LaurenSpiegel) [inline lib/services.js]: why does having a query.versionId indicate that the object is non-versioned?

Comment (ttvi) [inline lib/services.js]: No it's part of the opposite. This check is to see if we are deleting a non-versioned object or a specific version---cases when we need to physically delete object's data. The later part of the condition means when the object is versioning-enabled and the request is to delete a specific version; it is equivalent to `objectMD['x-amz-version-id'] !== 'null' && query.versionId`.

File: lib/api/bucketGet.js (hunk 4 above)
Comment (ttvi) [inline lib/api/bucketGet.js]: I was hesitate to do that because I did not know whether these attributes should be known locally to S3 only or should metadata also know about them.

File: lib/api/objectPut.js (hunk 9 above)
Comment (ttvi) [inline lib/api/objectPut.js]: This is another change in the S3-metadata communication. Now metadata is sending back `ETag` to S3. In the case of versioning, it needs to include the `versionId` as well.

File: lib/api/objectPut.js (hunk 8 above)
Comment (ttvi) [inline lib/api/objectPut.js]: Yes we should. In normal cases, `objMD['x-amz-version-id'] === 'null'`.

File: lib/routes/routePUT.js (hunk 10 above)
Comment (ttvi) [inline lib/routes/routePUT.js]: Right. It's the very first patch for versioning put to receive a `versionId` back after an `objectPut`. I will need to check with the call path and remove it.

File: lib/services.js (hunk 12 above)
Comment (ttvi) [inline lib/services.js]: I don't know what it is. WIll take a look.

File: lib/services.js (hunk 13 above)
Comment (ttvi) [inline lib/services.js]: Yes. It'd be better.

File: lib/api/bucketGet.js (hunk 3 above)
Comment (ttvi) [inline lib/api/bucketGet.js]: The `decodeVersion` might or might not need `JSON.parse`. In any case try/catch should be in the utils.

File: lib/api/bucketGet.js (hunk 5 above)
Comment (ttvi) [inline lib/api/bucketGet.js]: Yes, this will be changed.

File: lib/api/bucketGet.js (hunk 2 above)
Comment (ttvi) [inline lib/api/bucketGet.js]: Hmm, it should be in `metadataValParams`.

**Comment by ttvi** [comment]
Thanks @LaurenSpiegel , updates will be done soon.

File: lib/services.js (hunk 14 above)
Comment (LaurenSpiegel) [inline lib/services.js]: ok.  then, the log messages need to be updated.

File: package.json [hunk 15]
```diff
@@ -58,6 +58,7 @@
   },
   "scripts": {
     "ft_test": "mocha tester.js --compilers js:babel-core/register",
+    "test_aws": "babel-node tests/functional/aws-node-sdk/test/legacy/test.js",
```
Comment (LaurenSpiegel) [inline package.json]: I assume we don't want this in the final PR.  Note that this is not all of the node aws sdk tests.

File: tests/unit/api/objectPut.js [hunk 16]
```diff
@@ -201,6 +201,10 @@ describe('objectPut API', () => {
             });
     });
 
+    /**
+     * now overwriting an object in versioning means adding a new version, thus
+     * old data will node be deleted just be overwriting an existing object
```
Comment (LaurenSpiegel) [inline tests/unit/api/objectPut.js]: now overwriting an object in versioning means adding a new version, thus old data will not be deleted

File: package.json (hunk 15 above)
Comment (ttvi) [inline package.json]: oops

File: lib/api/bucketGet.js (hunk 2 above)
Comment (ttvi) [inline lib/api/bucketGet.js]: Actually I do use `versions` in `listParams` to inform metadata that this is a listing of all versions.

Comment (LaurenSpiegel) [inline lib/api/bucketGet.js]: ok.  Still would be clearer to just assign it to true explicitly.

**Comment by vrancurel** [comment]
We can merge that as there is no implication if we don't use versioning

**Comment by ttvi** [comment]
@vrancurel I'm separating the versioning and replication parts. I will call for review again after that.

**Comment by LaurenSpiegel** [comment]
@ttvi, would it be possible to keep the cb as the last argument in the functions if we are going to merge this?

**Comment by ttvi** [comment]
@LaurenSpiegel yes of course. But it will require searching for all usages and modify them. I will make it the last step of this PR.

**Comment by ttvi** [comment]
@ironman-machine please

File: lib/api/apiUtils/bucket/bucketDeletion.js [hunk 17]
```diff
@@ -110,7 +110,7 @@ export function deleteBucket(bucketMD, bucketName, canonicalID, log, cb) {
 
     return async.waterfall([
         function checkForObjectsStep(next) {
-            return metadata.listObject(bucketName, null, null, null, 1, log,
+            return metadata.listObject(bucketName, { maxKeys: 1 }, log,
```
Comment (MathieuCassagne) [inline lib/api/apiUtils/bucket/bucketDeletion.js]: Not really in the scope of the PR, but why do a listing of one key and not a get ?

File: lib/api/bucketPutVersioning.js [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+    const bucketName = request.bucketName;
+    const metadataValParams = {
+        authInfo,
+        bucketName,
+        requestType: 'bucketPutVersioning',
+        log,
+    };
+
+    return async.waterfall([
+        function waterfall1(next) {
```
Comment (MathieuCassagne) [inline lib/api/bucketPutVersioning.js]: Can you put an explicit name for the funciton in the waterfall ? :)

File: lib/api/apiUtils/bucket/bucketDeletion.js (hunk 17 above)
Comment (ttvi) [inline lib/api/apiUtils/bucket/bucketDeletion.js]: This is in the context of deleting a bucket. In order to delete a bucket, we need to ensure that it doesn't contain any object, and listing is a way to check that. I think using `maxKeys: 1` here is an optimization; you can also use the max threshold, but it just wastes disk ios and bandwidth.

File: lib/api/bucketPutVersioning.js [hunk 19]
```diff
@@ -0,0 +1,84 @@
+import async from 'async';
```
Comment (michael-zapata) [inline lib/api/bucketPutVersioning.js]: Nit: `import { waterfall } from 'async';` would work as well.

File: lib/api/bucketPutVersioning.js [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+
+    const bucketName = request.bucketName;
+    const metadataValParams = {
+        authInfo,
+        bucketName,
+        requestType: 'bucketPutVersioning',
+        log,
+    };
+
+    return async.waterfall([
```
Comment (michael-zapata) [inline lib/api/bucketPutVersioning.js]: Could we have each function defined out of the waterfall to avoid allocating it every time and not doing any v8 optimisation?

File: lib/metadata/wrapper.js [hunk 21]
```diff
@@ -117,9 +117,9 @@ const metadata = {
         });
     },
 
-    listObject: (bucketName, prefix, marker, delimiter, maxKeys, log, cb) => {
+    listObject: (bucketName, listingParams, log, cb) => {
```
Comment (michael-zapata) [inline lib/metadata/wrapper.js]: Why this change? The previous signature was explicit and avoided creating/allocating an object.

File: lib/routes/routePUT.js [hunk 22]
```diff
@@ -33,6 +33,42 @@ export default function routePUT(request, response, log, utapi) {
...
+                        errors.MalformedXML, null, response, null, log);
+                }
+                const xmlToParse = request.post;
+                return parseString(xmlToParse, (err, result) => {
+                    if (err) {
+                        log.debug('request xml is malformed');
+                        return routesUtils.responseNoBody(
+                            errors.MalformedXML, null, response, null, log);
+                    }
+                    const Status = result.VersioningConfiguration.Status ?
```
Comment (michael-zapata) [inline lib/routes/routePUT.js]: Nit: camelCasing would be welcome, the linter should complain about that.

File: lib/routes/routePUT.js [hunk 23]
```diff
@@ -33,6 +33,42 @@ export default function routePUT(request, response, log, utapi) {
...
+                        result.VersioningConfiguration.Status[0] : undefined;
+                    const MfaDelete = result.VersioningConfiguration.MfaDelete ?
+                        result.VersioningConfiguration.MfaDelete[0] : undefined;
+                    const validStatuses = ['Enabled', 'Suspended'];
```
Comment (michael-zapata) [inline lib/routes/routePUT.js]: Since those are defined once and for all, could we have them in the global scope instead of allocating them every time we go into this call? Also shouldn't this have `undefined` as well?

File: lib/routes/routePUT.js [hunk 24]
```diff
@@ -33,6 +33,42 @@ export default function routePUT(request, response, log, utapi) {
...
+                    const validMfaDeletes = [undefined, 'Enabled', 'Disabled'];
```
Comment (michael-zapata) [inline lib/routes/routePUT.js]: Same here.

File: tests/functional/aws-node-sdk/test/legacy/testBucketVersioning.js [hunk 25]
```diff
@@ -0,0 +1,131 @@
+import assert from 'assert';
```
Comment (michael-zapata) [inline tests/functional/aws-node-sdk/test/legacy/testBucketVersioning.js]: These aren't legacy tests, right?

Also, tests are not self-contained. (They should be able to run with an `.only()`)

File: lib/metadata/wrapper.js (hunk 21 above)
Comment (ttvi) [inline lib/metadata/wrapper.js]: For the ability to extend. Listing does not always take only these arguments, for example listing all versions of a bucket uses `max-keys`, `key-marker`, and `version-id-marker`.

File: lib/routes/routePUT.js (hunk 22 above)
Comment (ttvi) [inline lib/routes/routePUT.js]: It's AWS' format. See http://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketPUTVersioningStatus.html?shortFooter=true.

File: lib/routes/routePUT.js (hunk 23 above)
Comment (ttvi) [inline lib/routes/routePUT.js]: It will be at the global scope.
A valid `Status` should not be undefined while `MfaDelete` could be.

File: tests/functional/aws-node-sdk/test/legacy/testBucketVersioning.js (hunk 25 above)
Comment (ttvi) [inline tests/functional/aws-node-sdk/test/legacy/testBucketVersioning.js]: I'm not very familiar with the organization of the tests here.
Where do you think it should go?

File: lib/routes/routePUT.js (hunk 22 above)
Comment (michael-zapata) [inline lib/routes/routePUT.js]: Yeah, but we're talking about an internal variable here, hence my comment.

File: lib/metadata/wrapper.js (hunk 21 above)
Comment (michael-zapata) [inline lib/metadata/wrapper.js]: We should solve this by adding arguments as needed, then.

File: lib/routes/routePUT.js (hunk 22 above)
Comment (ttvi) [inline lib/routes/routePUT.js]: Oh right, the `Status` and `MfaDelete`. Thanks.

File: lib/metadata/wrapper.js (hunk 21 above)
Comment (ttvi) [inline lib/metadata/wrapper.js]: It would be a nightmare for readability. There is [List Objects Version 2](http://docs.aws.amazon.com/AmazonS3/latest/API/v2-RESTBucketGET.html?shortFooter=true) with its own set of parameters coming as well. I will keep this unless there are more requests like yours (consensus).

Comment (ttvi) [inline lib/metadata/wrapper.js]: Performance-wise, they are not different because inside this function, we have to create an object out of the listing parameters anyway.

**Comment by ttvi** [comment]
@MathieuCassagne @michael-zapata thanks for comments. I keep the revision in a new commit (to be squashed) to keep track of the changes.

File: tests/functional/aws-node-sdk/test/legacy/testBucketVersioning.js (hunk 25 above)
Comment (michael-zapata) [inline tests/functional/aws-node-sdk/test/legacy/testBucketVersioning.js]: We have `./tests/functional/aws-node-sdk/test/bucket` already created, you can add the file there.

File: lib/api/bucketPutVersioning.js [hunk 26]
```diff
@@ -16,6 +16,46 @@ import services from '../services';
...
+            });
+            return next(err);
+        }
+        return next(null, bucket);
+    });
+}
+
+function waterfall2ParseVersioningConfiguration(request, bucket, next) {
+    parseString(request.post, (err, result) => {
+        // just for linting; there should not be any parsing error here
```
Comment (michael-zapata) [inline lib/api/bucketPutVersioning.js]: Since this is based on user-input, you may very well have a parsing error (malicious intent or not).

File: lib/api/bucketPutVersioning.js [hunk 27]
```diff
@@ -16,6 +16,46 @@ import services from '../services';
     x-amz-mfa: [SerialNumber] [TokenCode]
  */
 
+
+function waterfall1ValidateAuthorization(mdValParams, log, next) {
```
Comment (michael-zapata) [inline lib/api/bucketPutVersioning.js]: I think you can skip the `waterfallX` here, it's a bit too verbose.

File: lib/api/bucketPutVersioning.js [hunk 28]
```diff
@@ -16,6 +16,46 @@ import services from '../services';
...
+    services.metadataValidateAuthorization(mdValParams, (err, bucket) => {
+        if (err) {
+            log.trace('request authorization failed', {
```
Comment (michael-zapata) [inline lib/api/bucketPutVersioning.js]: There's already logs for that, I don't think this is necessary. Especially since that would just get rid of that function since you could then use `services.metadataValidateAuthorization()` directly in the waterfall.

File: lib/api/bucketPutVersioning.js [hunk 29]
```diff
@@ -16,6 +16,46 @@ import services from '../services';
...
+                method: 'services.metadataValidateAuthorization',
+            });
+            return next(err);
+        }
+        return next(null, bucket);
+    });
+}
+
+function waterfall2ParseVersioningConfiguration(request, bucket, next) {
+    parseString(request.post, (err, result) => {
```
Comment (michael-zapata) [inline lib/api/bucketPutVersioning.js]: Nit: `parseString()` could also be in the waterfall itself. Nothing major there.

**Comment by ttvi** [comment]
Updated.

File: lib/api/bucketGetVersioning.js [hunk 30]
```diff
@@ -0,0 +1,67 @@
...
+ * bucketGetVersioning - Return Versioning Configuration for bucket
+ * @param {AuthInfo} authInfo - Instance of AuthInfo class with requester's info
+ * @param {object} request - http request object
+ * @param {object} log - Werelogs logger
+ * @param {function} callback - callback to respond to http request
+ *  with either error code or xml response body
+ * @return {undefined}
+ */
+export default function bucketGetVersioning(authInfo, request, log, callback) {
+    log.debug('processing request', { method: 'bucketGetACL' });
```
Comment (alexandre-merle) [inline lib/api/bucketGetVersioning.js]: BucketGetVersionning :)

File: lib/api/bucketPutVersioning.js [hunk 31]
```diff
@@ -0,0 +1,71 @@
...
+        log,
+    };
+
+    return waterfall([
+        next => services.metadataValidateAuthorization(metadataValParams,
+            (err, bucket) => next(err, bucket)), // to ignore null object
+        (bucket, next) => parseString(request.post, (err, result) => {
+            // just for linting; there should not be any parsing error here
+            if (err) {
+                return next(err);
```
Comment (alexandre-merle) [inline lib/api/bucketPutVersioning.js]: here its more a log of the error and send back an internal error :)

File: lib/metadata/wrapper.js (hunk 21 above)
Comment (alexandre-merle) [inline lib/metadata/wrapper.js]: i'm on @ttvi side, options should always be an object, to avoid modify everything when we add an option :)

File: lib/routes/routePUT.js [hunk 32]
```diff
@@ -33,6 +36,40 @@ export default function routePUT(request, response, log, utapi) {
...
+            } else if (request.query.versioning !== undefined) {
+                if (request.post === '') {
+                    log.debug('request xml is missing');
+                    return routesUtils.responseNoBody(
+                        errors.MalformedXML, null, response, null, log);
+                }
+                const xmlToParse = request.post;
+                return parseString(xmlToParse, (err, result) => {
+                    if (err) {
+                        log.debug('request xml is malformed');
```
Comment (alexandre-merle) [inline lib/routes/routePUT.js]: can you log the error here ? :) like err.stack :)

File: lib/api/bucketGetVersioning.js (hunk 30 above)
Comment (ttvi) [inline lib/api/bucketGetVersioning.js]: Thanks. Fixed.

File: lib/routes/routePUT.js (hunk 32 above)
Comment (ttvi) [inline lib/routes/routePUT.js]: I don't mind modifying it but don't you think the logging message is expressive enough?

File: lib/api/bucketPutVersioning.js (hunk 31 above)
Comment (ttvi) [inline lib/api/bucketPutVersioning.js]: The xml has actually been parsed here https://github.com/scality/S3/pull/91/files#diff-25f86c99d9d8badaa3509e13e8cc8edbR49.

File: lib/api/bucketPutVersioning.js [hunk 33]
```diff
@@ -0,0 +1,71 @@
...
+        bucketName,
+        requestType: 'bucketPutVersioning',
+        log,
+    };
+
+    return waterfall([
+        next => services.metadataValidateAuthorization(metadataValParams,
+            (err, bucket) => next(err, bucket)), // to ignore null object
+        (bucket, next) => parseString(request.post, (err, result) => {
+            // just for linting; there should not be any parsing error here
```
Comment (LaurenSpiegel) [inline lib/api/bucketPutVersioning.js]: I think you can remove this comment.

File: lib/api/bucketPutVersioning.js [hunk 34]
```diff
@@ -0,0 +1,71 @@
...
+            if (err) {
+                return next(err);
+            }
+            const versioningConfiguration = {};
+            if (result.VersioningConfiguration.Status) {
+                versioningConfiguration.Status =
+                    result.VersioningConfiguration.Status[0];
+            }
+            if (result.VersioningConfiguration.MfaDelete) {
```
Comment (LaurenSpiegel) [inline lib/api/bucketPutVersioning.js]: I think we should return a NotImplemented if this is in the request.  We do not want the user to think we are honoring this.

File: lib/api/bucketPutVersioning.js [hunk 35]
```diff
@@ -0,0 +1,71 @@
...
+ * @return {undefined}
+ */
+export default function bucketPutVersioning(authInfo, request, log, callback) {
+    log.debug('processing request', { method: 'bucketPutVersioning' });
+
+    const bucketName = request.bucketName;
+    const metadataValParams = {
+        authInfo,
+        bucketName,
+        requestType: 'bucketPutVersioning',
```
Comment (LaurenSpiegel) [inline lib/api/bucketPutVersioning.js]: This has to be addressed in the acl check functions.  The purpose of sending in the requestType is to know whether the requester has the authority under the acl's to perform the given action.  I don't think putBucketVersioning is addressed directly in the aws acl docs.  Perhaps only the bucket owner can call this api?

File: lib/api/bucketGetVersioning.js [hunk 36]
```diff
@@ -0,0 +1,67 @@
...
+ */
+export default function bucketGetVersioning(authInfo, request, log, callback) {
+    log.debug('processing request', { method: 'bucketGetVersioning' });
+
+    const bucketName = request.bucketName;
+
+    const metadataValParams = {
+        authInfo,
+        bucketName,
+        requestType: 'bucketGetVersioning',
```
Comment (LaurenSpiegel) [inline lib/api/bucketGetVersioning.js]: See comment regarding getBucketVersioning below and acl's.

File: lib/metadata/wrapper.js (hunk 21 above)
Comment (LaurenSpiegel) [inline lib/metadata/wrapper.js]: Vote for the @ttvi  approach. This way is so much more readable.

File: lib/routes/routeGET.js [hunk 37]
```diff
@@ -20,6 +20,14 @@ export default function routerGET(request, response, log, utapi) {
                     request.bucketName);
                 return routesUtils.responseXMLBody(err, xml, response, log);
             });
+        } else if (request.query.versioning !== undefined) {
+            api.callApiMethod('bucketGetVersioning', request, log,
+                (err, xml) => {
+                    // TODO push metrics for gucketGetVersioning
+                    // pushMetrics(err, log, utapi, 'bucketGetVersioning',
+                    //     request.bucketName);
```
Comment (LaurenSpiegel) [inline lib/routes/routeGET.js]: Please open an issue in s3 and utapi so this does not get forgotten.

File: lib/routes/routePUT.js [hunk 38]
```diff
@@ -33,6 +36,40 @@ export default function routePUT(request, response, log, utapi) {
...
                         log);
                 });
+            } else if (request.query.versioning !== undefined) {
+                if (request.post === '') {
+                    log.debug('request xml is missing');
+                    return routesUtils.responseNoBody(
+                        errors.MalformedXML, null, response, null, log);
+                }
+                const xmlToParse = request.post;
+                return parseString(xmlToParse, (err, result) => {
```
Comment (LaurenSpiegel) [inline lib/routes/routePUT.js]: Is there a reason to do the parsing here in the routes?  I'd prefer it in the api if possible, but it might be the right way to do things for error compatibility.  @nicolas2bert , do you have a view?

File: tests/functional/aws-node-sdk/test/bucket/testBucketVersioning.js [hunk 39]
```diff
@@ -0,0 +1,131 @@
...
+        const params = { Bucket: bucket };
+        // s3.getBucketVersioning(params, done);
+        s3.getBucketVersioning(params, (error, data) => {
+            assert.strictEqual(error, null);
+            assert.deepStrictEqual(data, { Status: 'Enabled' });
+            done();
+        });
+    });
+
+    // deletebucket test
```
Comment (LaurenSpiegel) [inline tests/functional/aws-node-sdk/test/bucket/testBucketVersioning.js]: you can put this into an after block rather than its own test.

File: tests/functional/aws-node-sdk/test/bucket/testBucketVersioning.js [hunk 40]
```diff
@@ -0,0 +1,131 @@
...
+
+    // setup test
+    before(() => {
+        const config = getConfig('default', { signatureVersion: 'v4' });
+
+        s3 = new S3(config);
+    });
+
+    // createbucket test
+    it('should create a bucket', done => {
```
Comment (LaurenSpiegel) [inline tests/functional/aws-node-sdk/test/bucket/testBucketVersioning.js]: probably makes more sense to put this in the before block rather than its own test.

File: lib/routes/routePUT.js [hunk 41]
```diff
@@ -33,6 +36,40 @@ export default function routePUT(request, response, log, utapi) {
...
+                    if (validStatuses.indexOf(status) < 0 ||
+                        validMfaDeletes.indexOf(mfaDelete) < 0) {
+                        log.debug('illegal versioning configuration');
+                        return routesUtils.responseNoBody(
+                            errors.IllegalVersioningConfigurationException,
+                            null, response, null, log);
+                    }
+                    return api.callApiMethod('bucketPutVersioning', request,
+                        log, err => {
+                            // TODO push metrics for bucketPutVersioning
```
Comment (LaurenSpiegel) [inline lib/routes/routePUT.js]: this could be in the same issue as the getBucketVersioning utapi integration.

File: lib/api/bucketPutVersioning.js [hunk 42]
```diff
@@ -0,0 +1,71 @@
...
+        log,
+    };
+
+    return waterfall([
+        next => services.metadataValidateAuthorization(metadataValParams,
+            (err, bucket) => next(err, bucket)), // to ignore null object
+        (bucket, next) => parseString(request.post, (err, result) => {
```
Comment (ttvi) [inline lib/api/bucketPutVersioning.js]: I think it's not feasible to make them static the way you want. Look at the full functions of the second waterfall:
```js
(bucket, cb) => parseString(req, (err, res) => {
    cb();
});
```
- The first anonymous function uses an extra variable `req` apart from the inputs so we cannot make it a static function unless we use `req` as a global variable.
- The second anonymous function uses the extra variable `next` as well so it's difficult to make it static.
You have any other solution?

And also, practically the function `bucketGetVersioning` should not be a frequent operation.

File: tests/functional/aws-node-sdk/test/bucket/testBucketVersioning.js (hunk 40 above)
Comment (ttvi) [inline tests/functional/aws-node-sdk/test/bucket/testBucketVersioning.js]: You have this.

File: tests/functional/aws-node-sdk/test/bucket/testBucketVersioning.js (hunk 39 above)
Comment (ttvi) [inline tests/functional/aws-node-sdk/test/bucket/testBucketVersioning.js]: and this.

File: lib/api/bucketPutVersioning.js (hunk 34 above)
Comment (ttvi) [inline lib/api/bucketPutVersioning.js]: I put it there, though the `NotImplemented` error will also confuse users whether versioning or only mfa deletion was not implemented.

File: lib/api/bucketPutVersioning.js (hunk 35 above)
Comment (ttvi) [inline lib/api/bucketPutVersioning.js]: Addressed.

File: lib/api/api.js [hunk 43]
```diff
@@ -87,9 +89,11 @@ const api = {
     bucketDelete,
     bucketGet,
     bucketGetACL,
+    bucketGetVersioning,
```
Comment (ttvi) [inline lib/api/api.js]: Mapped.

**Comment by ttvi** [comment]
All good. Waiting for https://github.com/scality/Arsenal/pull/170 to land.
Please don't make the comment count 100 :|.

**Comment by ttvi** [comment]
@ironman-machine

**Comment by ttvi** [comment]
CI failed because of using non-existing branch #170 of Arsenal. Retrying and passing here http://ci.ironmann.io/gh/scality/Integration/5116.

File: lib/api/api.js (hunk 43 above)
Comment (ttvi) [inline lib/api/api.js]: I don't think we need the integration test wrt policies because only the bucket owner can perform these actions.

**Comment by ttvi** [comment]
@LaurenSpiegel @michael-zapata your comments have been addressed/discussed. Let's finalize this.

File: lib/api/api.js (hunk 43 above)
Comment (LaurenSpiegel) [inline lib/api/api.js]: I think you mean the bucket owner can do so with respect to ACL's.  This doesn't mean that a user under the owner account does not have the right if a proper policy is attached.  I just tested with a user in real AWS and users can put and get bucket versioning.  That being said we can add an issue to integration to add this test after this PR.
